11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+						if (shaderWriteCacheFile)
+						{
+							system::IFile::success_t succ;
+							shaderWriteCacheFile->write(succ, serializedCache->getPointer(), 0, serializedCache->getSize());
+							if (!succ)
+								m_logger->log("Failed Writing To Shader Cache File.", ILogger::ELL_ERROR);
+						}
+					}
+					else
+						m_logger->log("Failed Creating Shader Cache File.", ILogger::ELL_ERROR);
+				}
+				else
+					m_logger->log("Failed Creating Shader Cache File.", ILogger::ELL_ERROR);
+			}
+
+		}
+		// Block and wait until kernel FFT is done before we drop the pipelines.
+		// Ideally not be lazy and create a latch that does nothing but capture the pipelines and gets called when scratch semaphore is signalled
+		const ISemaphore::SWaitInfo waitInfo = { m_timeline.get(), semaphorValue };
+
+		m_device->blockForSemaphores({ &waitInfo, 1 });
+
+		return true;
+	}
+
+	bool keepRunning() override { return m_keepRunning; }
+
+	// Right now it's one shot app, but I put this code here for easy refactoring when it refactors into it being a live app
+	void workLoopBody() override
+	{
+		uint32_t queueFamilyIndex = m_queue->getFamilyIndex();
+
+		// RESET COMMAND BUFFER OR SOMETHING
+
+		m_computeCmdBuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+		// Pipeline barrier: transition kernel spectrum images into read only, and outImage into general
+		IGPUCommandBuffer::SPipelineBarrierDependencyInfo imagePipelineBarrierInfo = {};
+		decltype(imagePipelineBarrierInfo)::image_barrier_t imgBarriers[CHANNELS + 1];
+		imagePipelineBarrierInfo.imgBarriers = { imgBarriers, CHANNELS + 1};
+
+		// outImage just needs a layout transition before it can be written to
+		imgBarriers[0].image = m_outImg.get();
+		imgBarriers[0].barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+		imgBarriers[0].barrier.dep.srcAccessMask = ACCESS_FLAGS::NONE;
+		imgBarriers[0].barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+		imgBarriers[0].barrier.dep.dstAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+		imgBarriers[0].oldLayout = IImage::LAYOUT::UNDEFINED;
+		imgBarriers[0].newLayout = IImage::LAYOUT::GENERAL;
+		imgBarriers[0].subresourceRange = { IGPUImage::EAF_COLOR_BIT, 0u, 1u, 0u, 1 };
+
+		// We need to wait on kernel spectrums to be written to before we can read them
+		for (auto i = 0u; i < CHANNELS; i++)
+		{
+			imgBarriers[i + 1].image = m_kernelNormalizedSpectrums[i]->getCreationParameters().image.get();
+			imgBarriers[i + 1].barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			imgBarriers[i + 1].barrier.dep.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+			imgBarriers[i + 1].barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			imgBarriers[i + 1].barrier.dep.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+			imgBarriers[i + 1].oldLayout = IImage::LAYOUT::GENERAL;
+			imgBarriers[i + 1].newLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
```

Comment:
the layout transition should have been done by the precomp dispatch

Reply:
with NONE destination masks (because no command between transition and semaphore signal)